File: app/src/main/java/com/jodelXposed/krokofant/features/UniqueDeviceIdentifierStuff.java
package com.jodelXposed.krokofant.features;

import com.jodelXposed.krokofant.utils.Settings;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.callbacks.XC_LoadPackage;
import org.json.JSONException;

import java.io.IOException;

import static com.jodelXposed.krokofant.utils.Log.xlog;
import static de.robv.android.xposed.XposedHelpers.findAndHookMethod;

public class UniqueDeviceIdentifierStuff {
    private static class UDI {
        static String GetUID = "Ar";
    }

    /**
     * Spoof UID
     */
    public UniqueDeviceIdentifierStuff(XC_LoadPackage.LoadPackageParam lpparam) {
        findAndHookMethod("com.jodelapp.jodelandroidv3.utilities.UniqueDeviceIdentifier", lpparam.classLoader, UDI.GetUID, new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                xlog("UDI = " + param.getResult());
                try {
                    Settings settings = Settings.getInstance();
                    if (!settings.isLoaded())
                        settings.load();
                    if (settings.getUid().length() == 0) {
                        settings.setUid((String) param.getResult());
                        settings.save();
                    } else if (settings.getUid().equals(param.getResult())) {
                        xlog("UDI not spoofed");
                    } else {
                        xlog("UDI spoof = " + settings.getUid());
                        param.setResult(settings.getUid());
                    }
                } catch (JSONException | IOException e) {
                    xlog("Error: " + e.getLocalizedMessage());
                }
            }
        });
    }
}

File: app/src/main/java/com/jodelXposed/hooks/udi/UniqueDeviceIdentifierStuff.kt
package com.jodelXposed.hooks.udi

import com.jodelXposed.hooks.helper.Log.dlog
import com.jodelXposed.utils.Options
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedHelpers.findAndHookMethod
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam

class UniqueDeviceIdentifierStuff(lpparam: LoadPackageParam) {
    init {
        try {
            findAndHookMethod(Options.hooks.Class_UniqueDeviceIdentifier, lpparam.classLoader, Options.hooks.Method_UDI_GetUdiMethod, object : XC_MethodHook() {
                @Throws(Throwable::class)
                override fun afterHookedMethod(param: MethodHookParam) {
                    if (Options.udi.active) {
                        val realUDI = param.result as String
                        if (Options.udi.udi.isEmpty()) {
                            Options.udi.udi = realUDI
                            dlog("Backing up UDI")
                            Options.save()
                            return
                        }

                        dlog("UDI spoof = ${Options.udi.udi}")
                        param.result = Options.udi.udi
                    }
                }
            })
        } catch (e: Exception) {
            e.printStackTrace()
            dlog("!!!!!!!!!! Failed loading UniqueDeviceIdentifier hook !!!!!!!!!!")
        }
    }
}
